Add tf2::BufferCoreInterface and tf2_ros::AsyncBufferInterface to use in tf2_ros::MessageFilter

jacobperron opened the pull request:
This work is towards enabling the use of MessageFilter in RViz (see https://github.com/ros2/rviz/pull/375#issuecomment-504124116).
The idea is to make RViz transformer plugins extend the new `tf2::FrameTransformerInterface` so that we can take advantage of `tf2::MessageFilter` with different transformer plugins (not just `tf2::BufferCore`).

This PR does three things:

1. Propose the new interface `tf2::FrameTransformerInterface` (533ccd6)
1. Make `tf2::BufferCore` implement the interface (f4c2c98)
1. Make `tf2::MessageFilter` use the interface (9550875)

jacobperron commented:
After an offline discussion with @wjwwood and @tfoote, I've done some refactoring.
Now, the new base class for `BufferCore` is named `BufferCoreInterface` for consistency.
The `waitForTransform` API has been moved to `tf2_ros::BufferInterface`, which is the ROS-aware (ie. time aware) interface on top of `BufferCore`.
I've also done a few other small things to make this PR slightly less intrusive to the existing code.

I still need to implement the timeout for `waitForTransform`.

jacobperron commented:
In order to implement the asynchronous timeout I've introduced the interface `CreateTimerInterface` that is registered with `tf2_ros::Buffer` after construction (@wjwwood as per our offline discussion).

I realized that `tf2::TransformListener` does not take a `tf2::Buffer` as an argument so, rather than changing it's constructor, I've left it alone and set the `CreateTimerInterface` in the test:

https://github.com/ros2/geometry2/blob/d75131301990d31aa664feb7dc3815a597da2a5f/tf2_ros/test/message_filter_test.cpp#L62-L64

wjwwood commented:
> I realized that tf2::TransformListener does not take a tf2::Buffer as an argument so, rather than changing it's constructor, I've left it alone and set the CreateTimerInterface in the test:

What? the `tf2_ros::TransformListener` takes the `tf2_ros::Buffer` as the first argument in the constructor. I don't entirely follow why the `tf2_ros::TransformListener` cannot call this method on the buffer? You said specifically `tf2` and not `tf2_ros` in your message, is that the difference?

jacobperron commented:
> What? the tf2_ros::TransformListener takes the tf2_ros::Buffer as the first argument in the constructor. I don't entirely follow why the tf2_ros::TransformListener cannot call this method on the buffer? You said specifically tf2 and not tf2_ros in your message, is that the difference?

I meant `tf2_ros::Buffer`. [`tf2_ros::TransformerListener`](https://github.com/ros2/geometry2/blob/d75131301990d31aa664feb7dc3815a597da2a5f/tf2_ros/include/tf2_ros/transform_listener.h#L54) accepts a `tf2::BufferCore` (not a `tf2_ros::Buffer`), so we can't register the new timer interface from inside the transform listener (unless we change the signature).

Edit: We can pass a `tf2_ros::Buffer` because it is extending `tf2::BufferCore`

wjwwood commented:
I see what you mean now. Oof. Does it make sense to move all the stuff (timeout based waitFor and the timer creation interface) into `tf2::Buffer`? I guess not huh?

I suppose we could add a new signature for `tf2_ros::TransformListener` that takes a `tf2_ros::Buffer` instead and only sets up the interface if that signature is used. I think most people will be passing an instance of the `tf2_ros::Buffer` anyways, and if not the new `waitFor` will not work unless they do so.

jacobperron commented:
Yeah, I don't think it makes sense to add the new async API to `tf2::BufferCore` since it is time agnostic.

Unless there's a use-case for passing a `tf2::BufferCore`, I would change the type passed to `tf2_ros::TransformListener` to `tf2_ros::Buffer`. But maybe it's safer to add a new signature as you suggest.

wjwwood commented:
Only if you wanted to backport it and not break API. Otherwise, unless @tfoote sees a different reason, I think changing it would be ok.

jacobperron commented:
I've already broken API by changing the signature of `tf2_ros::MessageFilter`, which I don't think is as straightforward to make backwards compatible; I'd have to think about it.

jacobperron commented:
I think this is ready for review. Here's a description of the changes:

* **Add pure virtual interface`tf2::BufferCoreInterface` and have `tf2::BufferCore` implement it.**
It is required that the object passed to `tf2_ros::MessageFilter` implements this interface.
The interface signature is very similar to the methods already in `tf2::BufferCore` to keep the diff smaller.

* **Add pure virtual interface `tf2_ros::AsyncBufferInterface` and have `tf2_ros::Buffer` implement it.**
The new interface declares a single method `waitForTransform`.
It is required that the object passed to `tf2_ros::MessageFilter` implements this interface.
Since `tf2` doesn't have a sense of time, `tf2_ros` seems to be the appropriate place to put the interface. An alternative approach is to add the async method to the existing `tf2_ros::BufferInterface`, but since `tf2_ros::MessageFilter` doesn't need any of the other interface methods, I chose to create a new interface. This cuts down on the number of methods that need to be implemented by third-parties (e.g. RViz).

* **Add pure virtual interface `tf2_ros::CreateTimerInterface` and concrete class `tf2_ros::CreateTimerROS`.**
Added a method `tf2_ros::Buffer::setCreateTimerInterface()` for registering the interface, which is required for computing timeouts for the new asynchronous method (`AsyncBufferInterface::waitForTransform()`).

* **`tf2_ros::MessageFilter` signature templated on the Buffer type.**
Static assertions are added to ensure the template type implements `tf2::BufferCoreInterface` and `tf2_ros::AsyncBufferInterface`.

Pending review and approval, I'll curate the commits so that they reflect the above description.

jacobperron commented:
See proposed usage in RViz: https://github.com/ros2/rviz/pull/422

tfoote commented on tf2_ros/src/buffer.cpp:
```diff
@@ -190,6 +191,73 @@ Buffer::canTransform(const std::string& target_frame, const tf2::TimePoint& targ
+                         const tf2::Duration& timeout, TransformReadyCallback callback)
+{
+  if (nullptr == timer_interface_) {
+    throw std::runtime_error("timer interface not set before call to waitForTransform");
```
This should be a custom Exception class. At least a generic tf2 Exception.

jacobperron replied:
Thanks for the review! I've added a custom exception in 7556ee5

tfoote reviewed (approved):
Overall this looks good to me. One small nitpick/suggestion on the exception type.

jacobperron commented:
Packages above tf2 and tf2_ros:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7523)](http://ci.ros2.org/job/ci_linux/7523/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3660)](http://ci.ros2.org/job/ci_linux-aarch64/3660/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6161)](http://ci.ros2.org/job/ci_osx/6161/)
* Windows [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7352)](https://ci.ros2.org/job/ci_windows/7352/)

Edit: Also need to build up to tf2_ros

wjwwood commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -88,7 +89,7 @@ static constexpr Duration BUFFER_CORE_DEFAULT_CACHE_TIME = std::chrono::seconds(
  * All function calls which pass frame ids can potentially throw the exception tf::LookupException
  */
-class BufferCore
+class BufferCore : public virtual BufferCoreInterface
```
@jacobperron and I spoke about this offline, and I think we can remove this virtual inheritance as well as a few other places in this pr and the rviz one. He's going to try to remove them and I'll continue my review after that. 👍

jacobperron replied:
Done 4160ac7

wjwwood commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -108,7 +109,7 @@ class BufferCore
   /** \brief Clear all data */
   TF2_PUBLIC
-  void clear();
+  void clear() override;
```
I'm also mildly concerned about making these functions virtual (though I don't see an alternative), because it will now require a vtable lookup, and these functions are called very frequently. Just bringing it up in case anyone else agrees, but I don't see an alternative at the moment as the standard way to avoid a vtable is CRTP and the way we're using this with plugins I think makes that option not viable.

jacobperron commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7535)](http://ci.ros2.org/job/ci_linux/7535/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3672)](http://ci.ros2.org/job/ci_linux-aarch64/3672/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6172)](http://ci.ros2.org/job/ci_osx/6172/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7364)](http://ci.ros2.org/job/ci_windows/7364/)

One of the new tests exposed an existing bug. A PR with a fix is open: https://github.com/ros2/geometry2/pull/127

I also made a couple other small fixes with this PR. @wjwwood PTAL

jacobperron commented:
I've rebased to get the fix from #127. The last three commits are new since the last review.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7547)](http://ci.ros2.org/job/ci_linux/7547/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3688)](http://ci.ros2.org/job/ci_linux-aarch64/3688/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6181)](http://ci.ros2.org/job/ci_osx/6181/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7373)](http://ci.ros2.org/job/ci_windows/7373/)

jacobperron committed:
Add pure virtual interface tf2::BufferCoreInterface

tf2::BufferCore implements this interface.
The plan is to use the interface in place of tf2::BufferCore (for example, in tf2_ros::MessageFilter).
This gives applications a way to provide custom implementations of the core TF interface.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add pure virtual interface tf2_ros::CreateTimerInterface

Add concrete class tf2_ros::CreateTimerROS that implements the new interface.
The interface will be used by tf2_ros::Buffer to implement asynchronous operations.
Providing a separate interface for interacting with timers lets us keep a separation of concerns
when it comes to executing user-defined callbacks.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add pure virtual interface tf2_ros::AsyncBufferInterface

The new interface declares a single method waitForTransform.
The interface is implemented by tf2_ros::Buffer, which makes use of tf2_ros::CreateTimerInterface.
This makes it easier for users to make asynchronous transform requests, rather than using the
methods provided by the base class tf2::BufferCore.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Template tf2_ros::MessageFilter on the buffer type

This change applications to use the message filter with custom buffer implementations.
The buffer type provided must implement tf2::BufferCoreInterface and tf2_ros::AsyncBufferInterface,
which is enforced with static assertions.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jdlangs commented:
@jacobperron this code is not compiling for me and I'm unable to figure out why. The trouble seems to be with the added `static_asserts` where my compiler can't find the `tf2::BufferCoreInterface` name, even though there's a clear header chain that includes `tf2/buffer_core_interface.h`:
```
[Processing: tf2_ros]
--- stderr: tf2_ros
In file included from /home/jlangsfeld/Workspaces/ros2_ws/src/geometry2/tf2_ros/test/message_filter_test.cpp:39:0:
/home/jlangsfeld/Workspaces/ros2_ws/src/geometry2/tf2_ros/include/tf2_ros/message_filter.h: In constructor ‘tf2_ros::MessageFilter<M, BufferT>::MessageFilter(BufferT&, const string&, uint32_t, const SharedPtr&, std::chrono::duration<_Rep, _Period>)’:
/home/jlangsfeld/Workspaces/ros2_wssrc/geometry2/tf2_ros/include/tf2_ros/message_filter.h:135:40: error: ‘BufferCoreInterface’ is not a member of ‘tf2’
     static_assert(std::is_base_of<tf2::BufferCoreInterface, BufferT>::value,
                                        ^~~~~~~~~~~~~~~~~~~
/home/jlangsfeld/Workspaces/ros2_ws/src/geometry2/tf2_ros/include/tf2_ros/message_filter.h:135:40: note: suggested alternative: ‘TimeCacheInterface’
     static_assert(std::is_base_of<tf2::BufferCoreInterface, BufferT>::value,
                                        ^~~~~~~~~~~~~~~~~~~
                                        TimeCacheInterface
```

If I explicitly add `#include <tf2/buffer_core_interface.h>` to `message_filter.h`, this error goes away, but the `static_assert` fails even though there's a clear inheritance chain from `tf2_ros::Buffer` to `tf2::BufferCoreInterface`:

```
--- stderr: tf2_ros
In file included from /home/jlangsfeld/Workspaces/ros2_ws/src/geometry2/tf2_ros/test/message_filter_test.cpp:39:0:
/home/jlangsfeld/Workspaces/ros2_ws/src/geometry2/tf2_ros/include/tf2_ros/message_filter.h: In instantiation of ‘tf2_ros::MessageFilter<M, BufferT>::MessageFilter(BufferT&, const string&, uint32_t, const SharedPtr&, std::chrono::duration<_Rep, _Period>) [with TimeRepT = long int; TimeT = std::ratio<1, 1000000000>; M = geometry_msgs::msg::PointStamped_<std::allocator<void> >; BufferT = tf2_ros::Buffer; std::__cxx11::string = std::__cxx11::basic_string<char>; uint32_t = unsigned int; rclcpp::Node::SharedPtr = std::shared_ptr<rclcpp::Node>]’:
/home/jlangsfeld/Workspaces/ros2_ws/src/geometry2/tf2_ros/test/message_filter_test.cpp:59:92:   required from here
/home/jlangsfeld/Workspaces/ros2_ws/src/geometry2/tf2_ros/include/tf2_ros/message_filter.h:136:5: error: static assertion failed: Buffer type must implement tf2::BufferCoreInterface
     static_assert(std::is_base_of<tf2::BufferCoreInterface, BufferT>::value,
```

Any suggestions?

jacobperron commented:
@jdlangs I can reproduce the error if I source a binary installation of Dashing and then try to compile geometry2. If I build ROS 2 from source, then I do not run into this issue. I suspect there's a mix-up happening with the Dashing installation, which doesn't contain the changes introduced in this PR, where tf2_ros is linking against the binary installation of tf2. ~~Unfortunately, I believe in order to get the tf2_ros library to link against the latest version of tf2, we also need to have source builds of its other direct dependencies (rclcpp, message_filters, and geometry_msgs). Because there have been API/ABI breaking changes to rclcpp and its dependencies since Dashing, I suggest building ROS 2 from source if you want to take advantage of this change.~~

Edit: fixed in #156

---

Since the message filter is referencing `tf2::BufferCoreInterface`, we better include the appropriate header in any case.

jacobperron commented:
@jdlangs Nevermind, this is a bug that can be resolved by changing the order of linked dependencies. See https://github.com/ros2/geometry2/pull/156 for a fix.
